spark/src/main/scala/io/delta/sharing/spark/DeltaSharingSource.scala

```diff
@@ -0,0 +1,566 @@
+  }
+
+  /**
+   * Returns the offset that starts from a specific delta table version. This function is
+   * called when starting a new stream query.
+   * @param fromVersion The version of the delta table to calculate the offset from.
+   * @param isStartingVersion - If true, will load fromVersion as a table snapshot(including files
+   *                            from previous versions). If false, will only load files since
+   *                            fromVersion.
+   * @param limits Indicates how much data can be processed by a micro batch.
+   */
+  protected def getStartingOffsetFromSpecificDeltaVersion(
+    fromVersion: Long,
+    isStartingVersion: Boolean,
+    limits: Option[AdmissionLimits]): Option[Offset] = {
+    val lastFileChange = getLastFileChangeWithRateLimit(
+      fromVersion,
+      fromIndex = -1L,
+      isStartingVersion = isStartingVersion,
+      limits)
+    if (lastFileChange.isEmpty) {
+      None
+    } else {
+      buildOffsetFromIndexedFile(lastFileChange.get, fromVersion, isStartingVersion)
+    }
+  }
+
+  /**
+   * Return the next offset when previous offset exists.
+   */
+  protected def getNextOffsetFromPreviousOffset(
+    previousOffset: DeltaSharingSourceOffset,
+    limits: Option[AdmissionLimits]): Option[Offset] = {
+    val lastFileChange = getLastFileChangeWithRateLimit(
+      previousOffset.tableVersion,
+      previousOffset.index,
+      previousOffset.isStartingVersion,
+      limits)
+
+    if (lastFileChange.isEmpty) {
+      // Return the previousOffset if there are no more changes, which still indicates until which
+      // offset we've processed the data.
+      Some(previousOffset)
+    } else {
+      buildOffsetFromIndexedFile(lastFileChange.get, previousOffset.tableVersion,
+        previousOffset.isStartingVersion)
+    }
+  }
+
+  /**
+   * Build the latest offset based on the last indexedFile. The function also checks if latest
+   * version is valid by comparing with previous version.
+   * @param lastIndexedFile - The last indexed file used to build offset from.
+   * @param previousOffsetVersion - Previous offset table version.
+   * @param ispreviousOffsetStartingVersion - Whether previous offset is starting version or not.
+   * @return the constructed offset.
+   */
+  private def buildOffsetFromIndexedFile(
+    lastIndexedFile: IndexedFile,
+    previousOffsetVersion: Long,
+    ispreviousOffsetStartingVersion: Boolean): Option[DeltaSharingSourceOffset] = {
+    val IndexedFile(v, i, _, isLastFileInVersion) = lastIndexedFile
+    assert(v >= previousOffsetVersion,
+      s"buildOffsetFromIndexedFile receives an invalid previousOffsetVersion: $v " +
+        s"(expected: >= $previousOffsetVersion), tableId: $tableId")
+
+    // If the last file in previous batch is the last file of that version, automatically bump
+    // to next version to skip accessing that version file altogether.
```

Comment:
It is not clear from the comment where do we skip accessing the version (it is when we call maybeGetFiles?)
Consider expanding on it.

Is it true that we are using the fact that logically <previousVersion, lastIndex> == <nextVersion, -1> ?
If so, perhaps mention it to make it clearer.